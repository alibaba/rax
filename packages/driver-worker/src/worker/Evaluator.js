export default class Evaluator {
  _taskPending = {};
  _id = 0;

  constructor(sender) {
    this.sender = sender;
  }

  apply(data) {
    const task = this._taskPending[data.id];
    // Task only can be resolved once.
    if (task) {
      const { resolve, reject } = task;
      if (data.type === 'success') {
        resolve(data.success);
      } else if (data.type === 'error') {
        reject(data.error);
      }

      this._taskPending[data.id] = null;
    }
  }

  _send(data) {
    this.sender({
      type: 'EvaluationRecord',
      data,
    });
  }

  /**
   * Query a variable's value.
   * get('location.href')
   */
  get(expression) {
    return this._eval({ code: expression });
  }

  /**
   * Assign value to an variable.
   * set('location.href', value)
   */
  set(expression, value) {
    return this._eval({ code: `${expression}=${value}` });
  }

  /**
   * Delete an variable.
   * delete('location.href')
   */
  delete(expression) {
    return this._eval({ code: `delete ${expression}` });
  }

  /**
   * Call a specfic method with params.
   * call('location.replace', '')
   */
  call(method) {
    var params = [];
    for (var l = arguments.length, i = 1; i < l; i++) {
      params[i - 1] = JSON.stringify(arguments[i]);
    }
    return this._eval({ code: `${method}(${params.toString()})` });
  }

  _eval({ code }) {
    const id = this._generateId();
    return new Promise((resolve, reject) => {
      this._taskPending[id] = { resolve, reject };
      this._send({ type: 'eval', id, code });
    });
  }

  /**
   * Generate a unique procedure id.
   */
  _generateId() {
    return this._id++;
  }
}
