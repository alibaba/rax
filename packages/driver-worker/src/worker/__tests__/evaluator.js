import Evaluator from '../Evaluator';

describe('Evaluator', () => {
  it('send', (done) => {
    const data = { hello: 'world' };
    const sender = (message) => {
      expect(message).toEqual({
        type: 'EvaluationRecord',
        data,
      });
      done();
    };
    const evaluator = new Evaluator(sender);
    evaluator._send(data);
  });

  it('get', (done) => {
    const sender = (message) => {
      expect(message).toMatchSnapshot();
      done();
    };
    const evaluator = new Evaluator(sender);
    evaluator.get('document.body.clientWidth');
  });

  it('set', (done) => {
    const sender = (message) => {
      expect(message).toMatchSnapshot();
      done();
    };
    const evaluator = new Evaluator(sender);
    evaluator.set('document.body.innerHTML', JSON.stringify('hello world'));
  });

  it('delete', (done) => {
    const sender = (message) => {
      expect(message).toMatchSnapshot();
      done();
    };
    const evaluator = new Evaluator(sender);
    evaluator.delete('foo.bar');
  });

  it('call with success', (done) => {
    const evaluator = new Evaluator((message) => {
      expect(message).toMatchSnapshot();

      setTimeout(() => {
        evaluator.apply({
          id: message.data.id,
          type: 'success',
          success: 'YES',
        });
      }, 300);
    });
    evaluator.call('foo.bar', 123, 'hello')
      .then((val) => {
        expect(val).toEqual('YES');
        done();
      });
  });

  it('call with error', (done) => {
    const evaluator = new Evaluator((message) => {
      expect(message).toMatchSnapshot();

      setTimeout(() => {
        evaluator.apply({
          id: message.data.id,
          type: 'error',
          error: 'NO',
        });
      }, 300);
    });
    evaluator.call('foo.bar', 123, 'hello')
      .catch((error) => {
        expect(error).toEqual('NO');
        done();
      });
  });
});
