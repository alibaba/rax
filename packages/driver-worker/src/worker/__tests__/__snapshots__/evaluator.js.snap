// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Evaluator call with error 1`] = `
Object {
  "data": Object {
    "code": "foo.bar(123,\\"hello\\")",
    "id": 0,
    "type": "eval",
  },
  "type": "EvaluationRecord",
}
`;

exports[`Evaluator call with success 1`] = `
Object {
  "data": Object {
    "code": "foo.bar(123,\\"hello\\")",
    "id": 0,
    "type": "eval",
  },
  "type": "EvaluationRecord",
}
`;

exports[`Evaluator delete 1`] = `
Object {
  "data": Object {
    "code": "delete foo.bar",
    "id": 0,
    "type": "eval",
  },
  "type": "EvaluationRecord",
}
`;

exports[`Evaluator get 1`] = `
Object {
  "data": Object {
    "code": "document.body.clientWidth",
    "id": 0,
    "type": "eval",
  },
  "type": "EvaluationRecord",
}
`;

exports[`Evaluator set 1`] = `
Object {
  "data": Object {
    "code": "document.body.innerHTML=\\"hello world\\"",
    "id": 0,
    "type": "eval",
  },
  "type": "EvaluationRecord",
}
`;
